src/client/lazy-app/Modal/index.tsx

```diff
@@ -0,0 +1,128 @@
+import { h, Component, VNode, Fragment } from 'preact';
+import * as style from './style.css';
+import 'add-css:./style.css';
+import { linkRef } from 'shared/prerendered-app/util';
+
+interface Props {}
+
+export interface ModalMessage {
+  icon: VNode;
+  title: string;
+  content: VNode;
+}
+
+interface State {
+  message: ModalMessage;
+  shown: boolean;
+}
+
+export default class Modal extends Component<Props, State> {
+  state: State = {
+    message: {
+      icon: <svg></svg>,
+      title: 'default',
+      content: <Fragment></Fragment>,
+    },
+    shown: false,
+  };
+
+  private dialogElement!: HTMLDialogElement;
+  static modalInstance?: Modal | undefined;
+
+  componentDidMount() {
+    // Once a transition ends, check if the modal should be closed (not just hidden)
+    // dialog.close() instantly hides the modal, so we call it AFTER fading it out i.e. on transition end
+    this.dialogElement.addEventListener(
+      'transitionend',
+      this._closeOnTransitionEnd.bind(this),
+    );
+    this.dialogElement.setAttribute('inert', 'enabled');
+
+    Modal.modalInstance = this;
+  }
+
+  private _closeOnTransitionEnd() {
+    // If modal does not exist
+    // Or if it's not being closed at the moment
+    if (
+      !this.dialogElement ||
+      !this.dialogElement.classList.contains(style.modalClosing)
+    )
+      return;
+
+    this.dialogElement.close();
+    this.dialogElement.classList.remove(style.modalClosing);
+    this.dialogElement.setAttribute('inert', 'enabled');
+  }
+
+  static showModal(message: ModalMessage) {
+    Modal.modalInstance?._showModal(message);
+  }
+
+  static hideModal() {
+    Modal.modalInstance?._hideModal();
+  }
+
+  private _showModal(message: ModalMessage) {
+    if (!this.dialogElement) return;
```

Comment:
It might be worth throwing here (something like “called `Modal.showModal()` without putting the `<Modal />` component being present”).

Reply:
Sure thing!